src/Compilers/CSharp/Portable/CSharpParseOptions.cs

```diff
@@ -122,19 +128,18 @@ public new CSharpParseOptions WithKind(SourceCodeKind kind)
 
         public CSharpParseOptions WithLanguageVersion(LanguageVersion version)
         {
-            version = version.MapLatestToVersion();
-
-            if (version == this.LanguageVersion)
+            if (version == this.SpecifiedLanguageVersion)
             {
                 return this;
             }
 
-            if (!version.IsValid())
+            var effectiveLanguageVersion = version.MapSpecifiedToEffectiveVersion();
+            if (!effectiveLanguageVersion.IsValid())
             {
                 throw new ArgumentOutOfRangeException(nameof(version));
             }
 
-            return new CSharpParseOptions(this) { LanguageVersion = version };
+            return new CSharpParseOptions(this) { SpecifiedLanguageVersion = version, LanguageVersion = effectiveLanguageVersion };
```

Comment:
Disagree.  Having a constructor argument for both values introduces a new error case to the system: the possibility specified and effective are incompatible.  The route chosen by this change is to give the user a single place to specify the version and the compiler will handle the mapping.

---
In reply to: [74496665](https://github.com/dotnet/roslyn/pull/12969#discussion_r74496665) [](ancestors = 74496665)